docs/types.md

```diff
@@ -0,0 +1,276 @@
+
+### Safety
+
+If your program survives the compiler's type solver (i.e., it compiles
+without a type error), it is likely to run correctly, often the very
+first time.  It's not that Irken programs don't crash, it's that they
+don't crash in the _way_ that C programs crash.  'segfaults' are rare
+with Irken, virtually always caused by either a bug in the compiler or
+an unsafe use of the Foreign Function Interface (calling out to C).
+
+For example, languages like C, C++, and Java all use the idea of a
+'NULL' pointer.  It's easy, even for a seasoned programmer, to forget
+to check a pointer for NULL.
+
+In Irken, there is no such beast: its equivalent is the `maybe` type.
+It captures the idea that a computation may or may not return a
+result.  Since Irken's pattern matching requires that every case is
+handled, every time, it's impossible to forget to handle the case
+where a return value is `(maybe:no)`.
+
+
+### Clarity of Thought
+
+This is hard to explain to someone unfamiliar with this kind of type
+system.  Irken (and ML/OCaml/Haskell) programming encourages a
+different approach to problem-solving.  Usually the first step is to
+think about the _types_, rather than the _algorithms_.  After many
+years of experience, I believe this is a 'better' way to program.  It
+often results in better code, sooner than the alternative approach of
+thinking about the algorithms first, and data structures last.
+
+Once you understand the problem well enough to have the data structures
+laid out, the algorithms usually fall into place.
+
+I've been programming for over 35 years, and I still experience a bit
+of shock and joy when a program works perfectly the first time I run
+it.
+
+## Type Inference
+
+The compiler will examine a program and try to solve all the types in
+the program, using the
+[Hindley-Milner](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system)
+algorithm.
+
+One of the goals of Irken is to reduce the use of type annotations.  I
+believe they are one of the major turn-offs of functional languages,
+especially for newcomers.  So I want Irken to _look_ like Scheme (or
+many other high level languages), but be something completely
+different underneath.
+
+Irken supports type annotations, and in some cases they are necessary
+for type safety, but for the most part they can be left off.
+
+Note: the compiler is capable of dumping all the types in a program if
+you need them for documentation or understanding. (give the compiler
+the `-types` option).
+
+### A simple example.
+
+Here's a factorial function:
+
+```scheme
+(define (fact n)
+  (if (= n 0)
+      1
+      (* n (fact (- n 1)))
+      ))
+```
+
+When the compiler walks this function, it begins with the innermost
+expressions.  For example, the `(= n 0)` expression contains lots of
+usable type information already.  We know that `0` is an `int`.  We
+know that `=` takes two `int`s as arguments and returns a `bool`.  We
```

Comment:
wait... are you saying that in irken `=` is boolean comparison operator? `==` elsewhere? That's at least worth pointing out. Also, why not `=?`, since "`a?`: a boolean function or variable - `(odd? n)`."